### ISSUE TITLE
Model / RND Network bonus

### ISSUE THREAD

**Issue by timruhkopf** [timruhkopf/AnytimeAcquisition#13]
This is where the theory hits the compiler. Building a Meta-Learned Optimizer using an Actor-Critic architecture with PPO and an RND exploration bonus requires carefully piecing together several distinct components.

To make this modular and clean, we will build this in PyTorch. The architecture will consist of a **History Encoder** (to project $x, y$ coordinates), a **Causal Transformer** (to process the sequence of evaluations), the **Actor/Critic Heads** (to predict the next $x$ and the expected AUC), and the **RND Module** (for the exploration bonus).

Here is the complete architectural implementation.

### 1. The Core Modules: Encoding and Exploration

First, we need the standard Positional Encoding and our Random Network Distillation (RND) module. The RND module will look at the newly sampled $x$ and output a surprise/novelty score.

```python
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.distributions import Normal
import math

class PositionalEncoding(nn.Module):
    def __init__(self, d_model: int, max_len: int = 5000):
        super().__init__()
        position = torch.arange(max_len).unsqueeze(1)
        div_term = torch.exp(torch.arange(0, d_model, 2) * (-math.log(10000.0) / d_model))
        pe = torch.zeros(max_len, 1, d_model)
        pe[:, 0, 0::2] = torch.sin(position * div_term)
        pe[:, 0, 1::2] = torch.cos(position * div_term)
        self.register_buffer('pe', pe)

    def forward(self, x):
        # x shape: [seq_len, batch_size, d_model]
        return x + self.pe[:x.size(0)]

class RNDBonus(nn.Module):
    """Random Network Distillation for Intrinsic Exploration Bonus"""
    def __init__(self, x_dim: int, latent_dim: int = 128):
        super().__init__()
        # Target network (Frozen)
        self.target = nn.Sequential(
            nn.Linear(x_dim, 256),
            nn.ReLU(),
            nn.Linear(256, latent_dim)
        )
        # Predictor network (Trainable)
        self.predictor = nn.Sequential(
            nn.Linear(x_dim, 256),
            nn.ReLU(),
            nn.Linear(256, 256),
            nn.ReLU(),
            nn.Linear(256, latent_dim)
        )
        
        # Freeze target network
        for param in self.target.parameters():
            param.requires_grad = False

    def forward(self, x):
        """Returns the MSE error between target and predictor as the bonus."""
        target_feature = self.target(x).detach()
        predicted_feature = self.predictor(x)
        intrinsic_reward = F.mse_loss(predicted_feature, target_feature, reduction='none').mean(dim=-1)
        return intrinsic_reward, predicted_feature, target_feature

```

### 2. The Master Agent: Transformer + Actor-Critic

Now we build the main `MetaBOAgent`.

Because you are doing Black-Box optimization, your variables are continuous. The **Actor** will output the parameters of a Multivariate Gaussian ($\mu$ and $\sigma$), from which we sample the next $x_{t+1}$. The **Critic** outputs the expected return (your anytime AUC).

```python
class MetaBOAgent(nn.Module):
    def __init__(self, x_dim: int, d_model: int = 128, n_heads: int = 4, 
                 num_layers: int = 3, max_seq_len: int = 1000):
        super().__init__()
        self.x_dim = x_dim
        self.d_model = d_model
        
        # 1. State Encoder: Projects (x, y) into the latent space
        # We assume y is a scalar, so input dim is x_dim + 1
        self.state_encoder = nn.Sequential(
            nn.Linear(x_dim + 1, d_model),
            nn.LayerNorm(d_model),
            nn.ReLU()
        )
        self.pos_encoder = PositionalEncoding(d_model, max_len=max_seq_len)
        
        # 2. Transformer Backbone (Causal)
        encoder_layers = nn.TransformerEncoderLayer(
            d_model=d_model, 
            nhead=n_heads, 
            dim_feedforward=d_model * 4,
            dropout=0.1,
            batch_first=True # Inputs are [batch, seq_len, d_model]
        )
        self.transformer = nn.TransformerEncoder(encoder_layers, num_layers)
        
        # 3. Actor Head (Policy)
        # Outputs mu and unnormalized log_std for continuous action space
        self.actor_mu = nn.Linear(d_model, x_dim)
        self.actor_log_std = nn.Parameter(torch.zeros(1, x_dim)) 
        
        # 4. Critic Head (Value / Reward Model)
        self.critic = nn.Sequential(
            nn.Linear(d_model, d_model),
            nn.ReLU(),
            nn.Linear(d_model, 1)
        )
        
        # 5. RND Module (Exploration)
        self.rnd = RNDBonus(x_dim=x_dim)

    def generate_causal_mask(self, seq_len: int, device: torch.device):
        """Prevents the model from looking into the future of the optimization trace."""
        mask = torch.triu(torch.ones(seq_len, seq_len, device=device) * float('-inf'), diagonal=1)
        return mask

    def forward(self, history_x, history_y):
        """
        history_x: [batch_size, seq_len, x_dim]
        history_y: [batch_size, seq_len, 1]
        """
        batch_size, seq_len, _ = history_x.shape
        device = history_x.device
        
        # Concatenate (x, y) pairs
        state = torch.cat([history_x, history_y], dim=-1) # [B, T, x_dim + 1]
        
        # Embed and add positional encoding
        embedded_state = self.state_encoder(state) # [B, T, d_model]
        
        # Transformer expects [seq_len, batch_size, d_model] if batch_first=False, 
        # but we set batch_first=True, so it remains [B, T, d_model]
        embedded_state = self.pos_encoder(embedded_state.transpose(0, 1)).transpose(0, 1)
        
        # Apply causal mask
        causal_mask = self.generate_causal_mask(seq_len, device)
        
        # Pass through Transformer
        transformer_out = self.transformer(embedded_state, mask=causal_mask, is_causal=True)
        
        # We only care about the representation of the *last* step to make our next decision
        last_step_features = transformer_out[:, -1, :] # [B, d_model]
        
        # -- ACTOR (Policy) --
        mu = self.actor_mu(last_step_features)
        # Often in BO, we squash mu to [0, 1] bounds using Sigmoid
        mu = torch.sigmoid(mu) 
        std = torch.exp(self.actor_log_std).expand_as(mu)
        
        dist = Normal(mu, std)
        
        # -- CRITIC (Value) --
        value = self.critic(last_step_features) # Predicts expected AUC
        
        return dist, value, transformer_out

    def get_action_and_value(self, history_x, history_y, action=None):
        """Used during the PPO rollout and update loop."""
        dist, value, _ = self.forward(history_x, history_y)
        
        if action is None:
            action = dist.sample() # Sample next x to evaluate
            
        log_prob = dist.log_prob(action).sum(axis=-1)
        entropy = dist.entropy().sum(axis=-1)
        
        return action, log_prob, entropy, value
    
    def get_intrinsic_reward(self, next_x):
        """Calculates the exploration bonus for sampling next_x."""
        intrinsic_reward, pred_feat, target_feat = self.rnd(next_x)
        return intrinsic_reward, pred_feat, target_feat

```

### Key Architectural Choices Explained:

1. **Causal Masking (`is_causal=True`)**: This is vital. When processing a sequence of 50 optimization steps, step 10 cannot know what happened at step 11. The mask ensures the model learns the trajectory sequentially.
2. **The Actor Head Squash (`torch.sigmoid`)**: Bayesian Optimization typically operates on a bounded search space (e.g., hyperparameters between $0$ and $1$). Using a Sigmoid activation on the $\mu$ output naturally bounds the policy's suggestions.
3. **Detaching the RND Target**: Notice `self.target(x).detach()` in the RND module. The target network acts as a randomized hashing function. If you accidentally let gradients flow into it, the predictor will easily collapse it, destroying your exploration bonus.

**Now that the engine is built, are you ready to dive into the PPO Training Loop script where we synchronize the rollout buffer, calculate the generalized advantage estimation (GAE), and apply the AUC reward shaping?**

Alright, let's wire up the brain of this beast. Building the architecture is only half the battle; the true magic of PPO lies in the training loop—specifically how you handle the rollout buffer, calculate the **Generalized Advantage Estimation (GAE)**, and clip the policy updates so the Transformer doesn't unlearn everything in a single bad batch.

For a Transformer-based PPO operating in continuous spaces, the optimizer and learning rate schedule are critical. Standard Adam will often fail here; you need **AdamW** coupled with a **Linear Warmup and Cosine Annealing** schedule to prevent the attention weights from destabilizing early on.

Here is the complete, detailed training engine to make your Meta-BO Agent learn.

---

## 1. The Rollout Buffer

First, we need a clean way to store the trajectories during the Bayesian Optimization rollouts.

```python
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader, TensorDataset

class RolloutBuffer:
    def __init__(self):
        self.history_x = []
        self.history_y = []
        self.actions = []
        self.logprobs = []
        self.rewards = []
        self.values = []
        self.is_terminals = []

    def clear(self):
        del self.history_x[:], self.history_y[:], self.actions[:]
        del self.logprobs[:], self.rewards[:], self.values[:], self.is_terminals[:]

```

---

## 2. The PPO Trainer Class

This class handles the environment interaction (sampling the synthetic prior), calculating the intrinsic/extrinsic rewards, computing GAE, and performing the PPO gradient updates.

```python
class PPOTrainer:
    def __init__(self, agent, lr=3e-4, gamma=0.99, lam=0.95, eps_clip=0.2, 
                 c_value=0.5, c_entropy=0.01, c_rnd=1.0, rnd_weight=0.1):
        self.agent = agent
        self.gamma = gamma       # Discount factor
        self.lam = lam           # GAE lambda
        self.eps_clip = eps_clip # PPO clipping parameter
        self.c_value = c_value   # Value loss coefficient
        self.c_entropy = c_entropy # Entropy bonus coefficient
        self.c_rnd = c_rnd       # RND predictor loss coefficient
        self.rnd_weight = rnd_weight # Weight of intrinsic reward
        
        # 1. Optimizers (AdamW is crucial for Transformers)
        # We use a separate, slightly higher LR for the RND predictor
        self.optimizer = optim.AdamW([
            {'params': self.agent.state_encoder.parameters()},
            {'params': self.agent.transformer.parameters()},
            {'params': self.agent.actor_mu.parameters()},
            {'params': self.agent.actor_log_std},
            {'params': self.agent.critic.parameters()}
        ], lr=lr, weight_decay=1e-4)
        
        self.rnd_optimizer = optim.Adam(self.agent.rnd.predictor.parameters(), lr=lr * 5)
        
        # 2. LR Scheduler (Warmup + Cosine Decay)
        self.scheduler = optim.lr_scheduler.CosineAnnealingWarmRestarts(
            self.optimizer, T_0=100, T_mult=2
        )
        
        self.buffer = RolloutBuffer()
        self.mse_loss = nn.MSELoss()

    def calculate_gae(self, rewards, values, is_terminals):
        """
        Calculates Generalized Advantage Estimation (GAE).
        """
        advantages = []
        returns = []
        gae = 0
        
        # Append 0 for the value of the terminal state
        values = torch.cat([values, torch.tensor([0.0]).to(values.device)])
        
        for step in reversed(range(len(rewards))):
            is_terminal = is_terminals[step]
            mask = 1.0 - is_terminal
            
            # TD Error: delta = r_t + gamma * V(s_{t+1}) - V(s_t)
            delta = rewards[step] + self.gamma * values[step + 1] * mask - values[step]
            
            # GAE: A_t = delta_t + gamma * lambda * A_{t+1}
            gae = delta + self.gamma * self.lam * mask * gae
            advantages.insert(0, gae)
            returns.insert(0, gae + values[step])
            
        return torch.stack(advantages), torch.stack(returns)

    def update(self, epochs=4, batch_size=64):
        # 1. Convert Buffer to Tensors
        old_history_x = torch.stack(self.buffer.history_x).detach()
        old_history_y = torch.stack(self.buffer.history_y).detach()
        old_actions = torch.stack(self.buffer.actions).detach()
        old_logprobs = torch.stack(self.buffer.logprobs).detach()
        rewards = torch.stack(self.buffer.rewards).detach()
        old_values = torch.stack(self.buffer.values).detach()
        is_terminals = torch.tensor(self.buffer.is_terminals, dtype=torch.float32)

        # 2. Calculate Advantages & Returns via GAE
        advantages, returns = self.calculate_gae(rewards, old_values, is_terminals)
        
        # Normalize advantages (Crucial for continuous control stability)
        advantages = (advantages - advantages.mean()) / (advantages.std() + 1e-8)

        # 3. Create DataLoader for Mini-batch Updates
        dataset = TensorDataset(old_history_x, old_history_y, old_actions, 
                                old_logprobs, advantages, returns)
        dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=True)

        

        # 4. PPO Update Loop
        for _ in range(epochs):
            for batch in dataloader:
                b_hist_x, b_hist_y, b_actions, b_old_logprobs, b_adv, b_returns = batch
                
                # Re-evaluate current policy
                dist, values, _ = self.agent(b_hist_x, b_hist_y)
                logprobs = dist.log_prob(b_actions).sum(axis=-1)
                dist_entropy = dist.entropy().sum(axis=-1)
                
                # -- ACTOR LOSS (Clipped Surrogate) --
                ratios = torch.exp(logprobs - b_old_logprobs)
                surr1 = ratios * b_adv
                surr2 = torch.clamp(ratios, 1 - self.eps_clip, 1 + self.eps_clip) * b_adv
                actor_loss = -torch.min(surr1, surr2).mean()
                
                # -- CRITIC LOSS --
                critic_loss = self.mse_loss(values.squeeze(), b_returns)
                
                # -- TOTAL PPO LOSS --
                loss = actor_loss + self.c_value * critic_loss - self.c_entropy * dist_entropy.mean()
                
                # Take Gradient Step for Transformer/Actor/Critic
                self.optimizer.zero_grad()
                loss.backward()
                torch.nn.utils.clip_grad_norm_(self.agent.parameters(), max_norm=0.5)
                self.optimizer.step()
                
                # -- RND LOSS (Train Predictor) --
                _, pred_feat, target_feat = self.agent.get_intrinsic_reward(b_actions)
                rnd_loss = self.mse_loss(pred_feat, target_feat.detach())
                
                self.rnd_optimizer.zero_grad()
                rnd_loss.backward()
                self.rnd_optimizer.step()

        # Step the learning rate scheduler
        self.scheduler.step()
        self.buffer.clear()

```

---

## 3. The Synthetic Rollout Loop (Putting it Together)

Here is how you actually collect the data using your "Anytime Incumbent" reward logic and the exploration bonus before calling `trainer.update()`.

```python
def train_meta_bo(agent, trainer, num_episodes=5000, max_bo_steps=50):
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    agent.to(device)
    
    for episode in range(num_episodes):
        # 1. Sample a new synthetic function f(x) from your GP Prior
        # (Assuming you have a function `sample_synthetic_function`)
        # f = sample_synthetic_function()
        
        # Initialize starting history (random initial point)
        current_x = torch.rand(1, 1, agent.x_dim).to(device)
        current_y = torch.tensor([[[0.0]]]).to(device) # Replace with f(current_x)
        
        best_y = current_y.item()
        
        for t in range(max_bo_steps):
            # 2. Get action from the Policy
            with torch.no_grad():
                action, logprob, _, value = agent.get_action_and_value(current_x, current_y)
                
                # Calculate Intrinsic Reward (Exploration Bonus)
                intrinsic_reward, _, _ = agent.get_intrinsic_reward(action)
            
            # 3. Evaluate the Environment
            # y_next = f(action)
            y_next = torch.tensor([[[torch.rand(1).item()]]]).to(device) # Mock evaluation
            
            # 4. Calculate Extrinsic Reward (Anytime Incumbent AUC logic)
            improvement = max(0.0, y_next.item() - best_y)
            if improvement > 0:
                best_y = y_next.item()
                extrinsic_reward = improvement * 10.0 # Scale up improvement
            else:
                extrinsic_reward = -0.01 # Small penalty for no improvement (Reward Shaping)
                
            # Total Reward
            total_reward = extrinsic_reward + trainer.rnd_weight * intrinsic_reward.item()
            
            # 5. Store in Buffer
            trainer.buffer.history_x.append(current_x.squeeze(0))
            trainer.buffer.history_y.append(current_y.squeeze(0))
            trainer.buffer.actions.append(action.squeeze(0))
            trainer.buffer.logprobs.append(logprob.squeeze(0))
            trainer.buffer.values.append(value.squeeze(0))
            trainer.buffer.rewards.append(torch.tensor(total_reward).to(device))
            trainer.buffer.is_terminals.append(t == max_bo_steps - 1)
            
            # 6. Update History for the next Transformer step
            current_x = torch.cat([current_x, action.unsqueeze(0)], dim=1)
            current_y = torch.cat([current_y, y_next], dim=1)
            
        # 7. Perform PPO Update every N episodes
        if (episode + 1) % 4 == 0: 
            trainer.update()
            print(f"Episode {episode+1} | Best Y Found: {best_y:.4f}")

```

### 3 "Gotchas" You Must Watch Out For:

1. **Observation Normalization**: In real BO, $y$ values can scale wildly. If you don't use a running normalization (like a Z-score) on the $y$ inputs before feeding them to the Transformer, the Value Network (Critic) will fail to converge.
2. **Action Standard Deviation**: Notice `self.agent.actor_log_std` in the architecture. This allows the model to *learn* its own uncertainty. If training collapses into a single point, force a minimum bound on it: `std = torch.clamp(std, min=1e-3)`.
3. **PPO Batching**: We defined the Transformer sequence shape as `[batch, seq_len, d_model]`. When PPO shuffles the dataloader, it is shuffling *entire sequences* of history, preserving the temporal masking.

**You now have the complete end-to-end engine. Would you like me to draft the specific Gaussian Process generation code (using `Botorch` or `GPyTorch`) to act as the synthetic environment for this training loop?**
